sled-agent/src/services.rs

```diff
@@ -2908,44 +2944,95 @@ impl ServiceManager {
 
                             if is_gimlet {
-                                // Ddm for a scrimlet needs to be configured to
-                                // talk to dendrite
-                                smfh.setprop("config/dpd_host", "[::1]")?;
-                                smfh.setprop("config/dpd_port", DENDRITE_PORT)?;
+                                mg_ddm_config = mg_ddm_config
+                                    .add_property(
+                                        "dpd_host", "astring", "[::1]",
+                                    )
+                                    .add_property(
+                                        "dpd_port",
+                                        "astring",
+                                        &DENDRITE_PORT.to_string(),
+                                    )
                             }
-                            smfh.setprop("config/dendrite", "true")?;
 
-                            smfh.refresh()?;
+                            mg_ddm_service = mg_ddm_service.add_instance(
+                                ServiceInstanceBuilder::new("default")
+                                    .add_property_group(mg_ddm_config),
+                            );
                         }
                     }
-
-                    debug!(self.inner.log, "enabling service");
-                    smfh.enable()?;
                 }
+
+                switch_zone_setup_service = switch_zone_setup_service
+                    .add_instance(
+                        ServiceInstanceBuilder::new("default")
+                            .add_property_group(switch_zone_setup_config),
+                    );
+
+                let profile = ProfileBuilder::new("omicron")
+                    .add_service(nw_setup_service)
+                    .add_service(disabled_dns_client_service)
+                    .add_service(mgs_service)
+                    .add_service(wicketd_service)
+                    .add_service(switch_zone_setup_service)
+                    .add_service(dendrite_service)
+                    .add_service(tfport_service)
+                    .add_service(lldpd_service)
+                    .add_service(pumpkind_service)
+                    .add_service(mgd_service)
+                    .add_service(mg_ddm_service)
+                    .add_service(uplink_service);
+                profile
+                    .add_to_zone(&self.inner.log, &installed_zone)
+                    .await
+                    .map_err(|err| {
+                        Error::io("Failed to setup Switch zone profile", err)
+                    })?;
             }
-        };
+        }
 
+        let running_zone = RunningZone::boot(installed_zone).await?;
+
+        // Part of the process to ensure bootstrap address is to set up
+        // an IPv6 address within the Global Zone.
+        // This means we cannot run bootstrap setup via a service running on
+        // the switch zone itself.
```

Comment:
Just to make sure I'm understanding this: These steps involve:

- calling `zlogin` to the switch zone, after it's booted
- setting up a route in the global zone, after the zone has been booted

Although I do think these steps are important, don't they make the switch zone pretty intractably not self-assembling?

Would it be possible to have the switch zone create this bootstrap address itself, and for the global zone to create this route before booting the switch zone?

Reply:
I spent quite a bit of time to get this working like you mention. Unfortunately, it's not precisely the route forwarding that's the problem. It's when you create the bootstrap address that you need to set up an IPv6 address on the Global zone.


[ensure_bootstrap_address()](https://github.com/oxidecomputer/omicron/blob/2f146485d582dee1e832bb3046d54b670a426f0a/illumos-utils/src/running_zone.rs#L625 ) -> [Zones::ensure_address](https://github.com/oxidecomputer/omicron/blob/main/illumos-utils/src/zone.rs#L499) -> [Zones::create_address](https://github.com/oxidecomputer/omicron/blob/main/illumos-utils/src/zone.rs#L808-L817). This create_address method contains the follwoing:

```rust
        // Do any prep work before allocating the address.
        //
        // Currently, this only happens when allocating IPv6 addresses in the
        // non-global zone - to access these addresses, we must first set up
        // an arbitrary IPv6 address within the Global Zone.
        if let Some(zone) = zone {
            match addrtype {
                AddressRequest::Dhcp => {}
                AddressRequest::Static(addr) => {
                    if addr.is_ipv6() {
                        // Finally, actually ensure that the v6 address we want
                        // exists within the zone.
                        let link_local_addrobj =
                            addrobj.link_local_on_same_interface()?;
                        Self::ensure_has_link_local_v6_address(
                            Some(zone),
                            &link_local_addrobj,
                        )?;
                    }
                }
            }
        };
```

> Although I do think these steps are important, don't they make the switch zone pretty intractably not self-assembling?

Perhaps? :(

Reply:
Ah, if that's the issue, couldn't we allocate all the "state that needs to exist in the global zone" before we boot the switch zone?

E.g.:

1. Pre-Boot:
- GZ ensure an arbitrary link local v6 address exists
- GZ creates the route necessary to access the switch zone bootstrap address
- GZ edits the manifest of the switch zone, instructing it to create a specific bootstrap address

2. Switch Zone Boots
- Switch zones creates bootstrap address

This way, if the switch zone has booted, the sled agent can safely assume all this routing exists. If we don't keep this order strict, then "after an arbitrary reboot" the sled agent doesn't really know if the switch zone has been partially or fully initialized.

Reply:
I'll give it a go again, see if I can make this work

Reply:
Update: I have moved setting the link local address and bootstrap address to the switch zone setup service. Sadly, forwarding bootstrap traffic cannot be done before the zone boots :( There are a few reasons for this:

- Forwarding traffic won't work until the interface is up.
- The bootstrap interface is set up after the switch zone starts.
- We don't know what the interface will be called beforehand. The switch zone tries to initialize on a loop until it's able to locate "gzonly.txt" in any baseline directory (this is a host OS thing). Each time the zone attempts to start, the bootstrap interface is given a different name (oxBootstrap0, oxBootstrap1, oxBootstrap2 etc). This means we can't even do something hacky like have the `route add` command running on a loop in the background before boot because we don't know which boot will succeed, and therefore what name the interface will have.

Reply:
but hey, at least we're not calling `zlogin` anymore? :D

Reply:
I'm trying to make this comment make sense relative to what I'm seeing in the code -- and in the most recent revision of this PR, I see:

- A call which boots the zone:

```rust
let running_zone = RunningZone::boot(installed_zone).await?;
```

- Then, a retry loop which tries to call:

```rust
running_zone
  .add_bootstrap_route(
    BOOTSTRAP_PREFIX,
    self.inner.global_zone_bootstrap_link_local_address,
    bootstrap_name
  )
  ...
```

The invocation to `add_bootstrap_route` is here:

https://github.com/karencfv/omicron/blob/64b288d94d6f3128be6d03764429e23d242a8b02/illumos-utils/src/running_zone.rs#L782-L798

And it effectively is using `zlogin` (technically, it's using `zone_enter`, but it's functionally doing the same thing -- running a command within the switch zone, on behalf of the sled agent).

This code to add the bootstrap route:
- Runs in the switch zone
- Relies on inputs that appear known before the switch zone boots (BOOTSTRAP_PREFIX is a constant, `global_zone_bootstrap_link_local_address` is known before the zone boots, and `bootstrap_name` is derived from `bootstrap_name_and_address`, which is known at the start of the `initialize_zone` function).
- Is invoked by the sled agent

This is the most critical bit I'm trying to poke at: **after the sled agent boots the zone, I want to minimize -- and ideally fully eliminate -- any subsequent setup that needs to happen via zlogin / zone_enter**.

Reply:
After a quick offline chat with Sean and Ry, none of us can remember why they switch zone has a bootstrap address at all.  I can see `wicket` listening on it, but AFAIK we always talk to `wicket` using a link-local address.  It's possible that its existence is vestigial.  It may also have been added in anticipation of a need, which never came to fruition.

In any case,  I think it's worth skipping this part of the switch zone setup to see what (if anything) breaks.

Reply:
> And it effectively is using zlogin (technically, it's using zone_enter, but it's functionally doing the same thing

Ha! I was under the impression the command to forward bootstrap traffic was running/had to run from the global zone. I added some temporary debug logging in commit [9499512](https://github.com/oxidecomputer/omicron/pull/5593/commits/9499512dd67d4daed2445083ffd5c8378278ce91) to see exactly what commands were being run. Adding the bootstrap address and link local had `pfexec zlogin oxz_switch` while the traffic forwarding command did not:

```console
root@g3:~# cat $(svcs -L sled-agent) | grep "DEBUG" | grep "cmd" | looker
02:39:22.169Z INFO SledAgent (ServiceManager): DEBUG Zones::ensure_has_link_local_v6_address: No link-local address was found,
                attempt to make one.
    cmd = Command {\n    program: "/usr/bin/pfexec",\n    args: [\n        "/usr/bin/pfexec",\n        "/usr/sbin/zlogin",\n        "oxz_switch",\n        "/usr/sbin/ipadm",\n        "create-addr",\n        "-t",\n        "-T",\n        "addrconf",\n        "oxBootstrap36/ll",\n    ],\n}
    file = illumos-utils/src/zone.rs:792
    zone = oxz_switch
02:39:26.196Z INFO SledAgent (ServiceManager): DEBUG Zones::create_address_internal: Attempt to create the requested address
    addrobj = AddrObject {\n    interface: "oxBootstrap36",\n    name: "bootstrap6",\n}
    addrtype = Static(\n    V6(\n        Ipv6Network {\n            addr: fdb0:a840:2500:7::2,\n            prefix: 64,\n        },\n    ),\n)
    cmd = Command {\n    program: "/usr/bin/pfexec",\n    args: [\n        "/usr/bin/pfexec",\n        "/usr/sbin/zlogin",\n        "oxz_switch",\n        "/usr/sbin/ipadm",\n        "create-addr",\n        "-t",\n        "-T",\n        "static",\n        "-a",\n        "fdb0:a840:2500:7::2/64",\n        "oxBootstrap36/bootstrap6",\n    ],\n}
    file = illumos-utils/src/zone.rs:713
    zone = Some(\n    "oxz_switch",\n)
02:39:28.719Z INFO SledAgent (ServiceManager): DEBUG RunningZone::add_bootstrap_route: Adding bootstrap route
    bootstrap_prefix = 64944
    cmd = [\n    "/usr/sbin/route",\n    "add",\n    "-inet6",\n    "fdb0::/16",\n    "fe80::8:20ff:fe6d:b0da",\n    "-ifp",\n    "oxBootstrap36",\n]
    file = illumos-utils/src/running_zone.rs:805
    gz_bootstrap_addr = fe80::8:20ff:fe6d:b0da
    zone_vnic_name = "oxBootstrap36"
```

I have added this command to the switch_zone_setup service and so far it works fine on the a4x2 testbed. I will test on Madrid once the TUF repo is ready.

> After a quick offline chat with Sean and Ry, none of us can remember why they switch zone has a bootstrap address at all. I can see wicket listening on it, but AFAIK we always talk to wicket using a link-local address. It's possible that its existence is vestigial. It may also have been added in anticipation of a need, which never came to fruition.

> In any case, I think it's worth skipping this part of the switch zone setup to see what (if anything) breaks.

Good to know! After this PR is merged, there are a few bits of the switch zone startup flow I was planning to clean up in a follow up PR. Is it OK with you if I attempt to remove the bootstrap address in that follow up PR? This one has been open for quite a bit now, and I'd like to minimise the risk of things breaking :sweat_smile:

Reply:
Sure.  I was only offering this as a way to put this PR to bed more quickly.  If it will actually slow you down, then it's totally fine to address it (heh) in a follow-on.

Reply:
Thanks!

Reply:
> After a quick offline chat with Sean and Ry, none of us can remember why they switch zone has a bootstrap address at all. I can see `wicket` listening on it, but AFAIK we always talk to `wicket` using a link-local address. It's possible that its existence is vestigial. It may also have been added in anticipation of a need, which never came to fruition.
> 
> In any case, I think it's worth skipping this part of the switch zone setup to see what (if anything) breaks.

Sorry for the delay in reading this. Busy week :) I added this a long time ago. The original issue is [here](https://github.com/oxidecomputer/omicron/issues/2298), with a somewhat lengthy discussion. 

The bootstrap address in the switch zone is needed so that installinator can pull the real OS phase 2 and Control Plane images over the bootstrap network. Installinator [uses ddmd] (https://github.com/oxidecomputer/omicron/blob/main/installinator/src/peers.rs#L50-L85) to find wicketd on the bootstrap network so that it can pull artifacts from it.

Unfortunately, since this is a mupdate related issue, it's possible, even likely, that if your removed this bootstrap network related code CI would still pass and a mupdate of madrid to *this* new code would work. However, mupdating again likely would not work, because installinator can't retrieve from wicketd and I don't believe we actually serve artifacts from sled-agent yet. This would also cause any new installs after this code was merged to fail. 

In short, I'm glad you didn't remove the bootstrap address, and please don't do it in a follow up :). This is yet another reason we really need automated tests for mupdate. Until then, Chesterton's fence y'all!

Reply:
Oof :sweat_smile: